cpp/src/phonenumbers/phonenumberutil.cc

```diff
@@ -1527,12 +1525,76 @@ void PhoneNumberUtil::FormatNsnWithCarrier(const string& number,
       : metadata.intl_number_format();
   const NumberFormat* formatting_pattern =
       ChooseFormattingPatternForNumber(available_formats, number);
+  NumberFormat new_format;
+  bool consistent_optional_rule = TRUE;
+  bool consistent_fg_rule = TRUE;
   if (!formatting_pattern) {
-    formatted_number->assign(number);
-  } else {
-    FormatNsnUsingPatternWithCarrier(number, *formatting_pattern, number_format,
-                                     carrier_code, formatted_number);
+      // Set the pattern to the possible national patterns, check the number
+      // matches this or really we shouldn't be doing this.
+      const RegExp& national_number_pattern =
+          reg_exps_->regexp_cache_->GetRegExp(
+              StrCat("(",
+                     metadata.general_desc().national_number_pattern(),
+                     ")"));
+      std::cout << "Matching pattern: " << StrCat("(",
+                     metadata.general_desc().national_number_pattern(),
+                     ")") << "\n";
+      if (TestNumberLengthAgainstPattern(national_number_pattern,
+                                         number) != IS_POSSIBLE)
+      {
+        formatted_number->assign(number);
+        std::cout << "Fail out." << "\n";
+        return;
+      }
+      new_format.set_pattern(
```

Comment:
Explain a bit what you're doing here - "set up a new format to apply sensible defaults, rather than just doing no formatting on the number" - in fact this might be a slightly higer-level comment, you decide where it goes best!